File: mycollab-services/src/main/java/com/esofthead/mycollab/module/crm/service/ibatis/OpportunityServiceImpl.java
/**
 * This file is part of mycollab-services.
 *
 * mycollab-services is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mycollab-services is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mycollab-services.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.esofthead.mycollab.module.crm.service.ibatis;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.esofthead.mycollab.common.ModuleNameConstants;
import com.esofthead.mycollab.common.domain.GroupItem;
import com.esofthead.mycollab.common.interceptor.aspect.Auditable;
import com.esofthead.mycollab.common.interceptor.aspect.Traceable;
import com.esofthead.mycollab.common.interceptor.aspect.Watchable;
import com.esofthead.mycollab.core.cache.CacheKey;
import com.esofthead.mycollab.core.persistence.ICrudGenericDAO;
import com.esofthead.mycollab.core.persistence.ISearchableDAO;
import com.esofthead.mycollab.core.persistence.service.DefaultService;
import com.esofthead.mycollab.module.crm.CrmTypeConstants;
import com.esofthead.mycollab.module.crm.dao.OpportunityLeadMapper;
import com.esofthead.mycollab.module.crm.dao.OpportunityMapper;
import com.esofthead.mycollab.module.crm.dao.OpportunityMapperExt;
import com.esofthead.mycollab.module.crm.domain.Opportunity;
import com.esofthead.mycollab.module.crm.domain.OpportunityLead;
import com.esofthead.mycollab.module.crm.domain.OpportunityLeadExample;
import com.esofthead.mycollab.module.crm.domain.SimpleOpportunity;
import com.esofthead.mycollab.module.crm.domain.criteria.OpportunitySearchCriteria;
import com.esofthead.mycollab.module.crm.service.OpportunityService;
import com.esofthead.mycollab.schedule.email.crm.OpportunityRelayEmailNotificationAction;

/**
 * 
 * @author MyCollab Ltd.
 * @since 1.0
 * 
 */
@Service
@Transactional
@Traceable(module = ModuleNameConstants.CRM, type = CrmTypeConstants.OPPORTUNITY, nameField = "opportunityname")
@Auditable(module = ModuleNameConstants.CRM, type = CrmTypeConstants.OPPORTUNITY)
@Watchable(type = CrmTypeConstants.OPPORTUNITY, userFieldName = "assignuser", emailHandlerBean = OpportunityRelayEmailNotificationAction.class)
public class OpportunityServiceImpl extends
		DefaultService<Integer, Opportunity, OpportunitySearchCriteria>
		implements OpportunityService {

	@Autowired
	private OpportunityMapper opportunityMapper;
	@Autowired
	private OpportunityMapperExt opportunityMapperExt;
	@Autowired
	private OpportunityLeadMapper opportunityLeadMapper;

	@Override
	public ICrudGenericDAO<Integer, Opportunity> getCrudMapper() {
		return opportunityMapper;
	}

	@Override
	public ISearchableDAO<OpportunitySearchCriteria> getSearchMapper() {
		return opportunityMapperExt;
	}

	@Override
	public SimpleOpportunity findById(int opportunityId, int sAccountId) {
		return opportunityMapperExt.findById(opportunityId);
	}

	@Override
	public List<GroupItem> getSalesStageSummary(
			OpportunitySearchCriteria criteria) {
		return opportunityMapperExt.getSalesStageSummary(criteria);
	}

	@Override
	public List<GroupItem> getLeadSourcesSummary(
			OpportunitySearchCriteria criteria) {
		return opportunityMapperExt.getLeadSourcesSummary(criteria);
	}

	@Override
	public List<GroupItem> getPipeline(
			@CacheKey OpportunitySearchCriteria criteria) {
		return opportunityMapperExt.getPipeline(criteria);
	}

	@Override
	public void saveOpportunityLeadRelationship(
			List<OpportunityLead> associateLeads, Integer sAccountId) {
		for (OpportunityLead associateLead : associateLeads) {
			OpportunityLeadExample ex = new OpportunityLeadExample();
			ex.createCriteria()
					.andOpportunityidEqualTo(associateLead.getOpportunityid())
					.andLeadidEqualTo(associateLead.getLeadid());
			if (opportunityLeadMapper.countByExample(ex) == 0) {
				opportunityLeadMapper.insert(associateLead);
			}
		}
	}

	@Override
	public void removeOpportunityLeadRelationship(
			OpportunityLead associateLead, Integer sAccountId) {
		OpportunityLeadExample ex = new OpportunityLeadExample();
		ex.createCriteria()
				.andOpportunityidEqualTo(associateLead.getOpportunityid())
				.andLeadidEqualTo(associateLead.getLeadid());
		opportunityLeadMapper.deleteByExample(ex);
	}

	@Override
	public SimpleOpportunity findOpportunityAssoWithConvertedLead(int leadId,
			@CacheKey int accountId) {
		return opportunityMapperExt
				.findOpportunityAssoWithConvertedLead(leadId);
	}
}

File: mycollab-services/src/main/java/com/mycollab/module/crm/service/impl/OpportunityServiceImpl.kt
/**
 * mycollab-services - Parent pom providing dependency and plugin management for applications
		built with Maven
 * Copyright © 2017 MyCollab (support@mycollab.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
/**
 * mycollab-services - Parent pom providing dependency and plugin management for applications
		built with Maven
 * Copyright © ${project.inceptionYear} MyCollab (support@mycollab.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package com.mycollab.module.crm.service.impl

import com.mycollab.aspect.ClassInfo
import com.mycollab.aspect.ClassInfoMap
import com.mycollab.aspect.Traceable
import com.mycollab.aspect.Watchable
import com.mycollab.common.ModuleNameConstants
import com.mycollab.common.domain.GroupItem
import com.mycollab.core.cache.CacheKey
import com.mycollab.db.persistence.ICrudGenericDAO
import com.mycollab.db.persistence.ISearchableDAO
import com.mycollab.db.persistence.service.DefaultService
import com.mycollab.module.crm.CrmTypeConstants
import com.mycollab.module.crm.dao.OpportunityLeadMapper
import com.mycollab.module.crm.dao.OpportunityMapper
import com.mycollab.module.crm.dao.OpportunityMapperExt
import com.mycollab.module.crm.domain.*
import com.mycollab.module.crm.domain.criteria.OpportunitySearchCriteria
import com.mycollab.module.crm.service.ContactService
import com.mycollab.module.crm.service.OpportunityService
import com.mycollab.spring.AppContextUtil
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

/**
 * @author MyCollab Ltd.
 * @since 1.0
 */
@Service
@Transactional
@Traceable(nameField = "opportunityname")
@Watchable(userFieldName = "assignuser")
class OpportunityServiceImpl(private val opportunityMapper: OpportunityMapper,
                             private val opportunityMapperExt: OpportunityMapperExt,
                             private val opportunityLeadMapper: OpportunityLeadMapper) : DefaultService<Int, Opportunity, OpportunitySearchCriteria>(), OpportunityService {

    override val crudMapper: ICrudGenericDAO<Int, Opportunity>
        get() = opportunityMapper as ICrudGenericDAO<Int, Opportunity>

    override val searchMapper: ISearchableDAO<OpportunitySearchCriteria>
        get() = opportunityMapperExt

    override fun findById(opportunityId: Int, sAccountId: Int): SimpleOpportunity? {
        return opportunityMapperExt!!.findById(opportunityId)
    }

    override fun saveWithSession(opportunity: Opportunity, username: String?): Int {
        val result = super.saveWithSession(opportunity, username)
        if (opportunity.extraData != null && opportunity.extraData is SimpleContact) {
            val associateOpportunity = ContactOpportunity()
            associateOpportunity.opportunityid = opportunity.id
            associateOpportunity.contactid = (opportunity.extraData as SimpleContact).id
            associateOpportunity.createdtime = GregorianCalendar().time
            val contactService = AppContextUtil.getSpringBean(ContactService::class.java)
            contactService.saveContactOpportunityRelationship(listOf(associateOpportunity),
                    opportunity.saccountid)
        }
        return result
    }

    override fun getSalesStageSummary(criteria: OpportunitySearchCriteria): List<GroupItem> {
        return opportunityMapperExt.getSalesStageSummary(criteria)
    }

    override fun getLeadSourcesSummary(criteria: OpportunitySearchCriteria): List<GroupItem> {
        return opportunityMapperExt.getLeadSourcesSummary(criteria)
    }

    override fun getPipeline(@CacheKey criteria: OpportunitySearchCriteria): List<GroupItem> {
        return opportunityMapperExt.getPipeline(criteria)
    }

    override fun saveOpportunityLeadRelationship(associateLeads: List<OpportunityLead>, sAccountId: Int?) {
        for (associateLead in associateLeads) {
            val ex = OpportunityLeadExample()
            ex.createCriteria().andOpportunityidEqualTo(associateLead.opportunityid)
                    .andLeadidEqualTo(associateLead.leadid)
            if (opportunityLeadMapper.countByExample(ex) == 0L) {
                opportunityLeadMapper.insert(associateLead)
            }
        }
    }

    override fun removeOpportunityLeadRelationship(associateLead: OpportunityLead, sAccountId: Int?) {
        val ex = OpportunityLeadExample()
        ex.createCriteria().andOpportunityidEqualTo(associateLead.opportunityid)
                .andLeadidEqualTo(associateLead.leadid)
        opportunityLeadMapper.deleteByExample(ex)
    }

    override fun findOpportunityAssoWithConvertedLead(leadId: Int?, @CacheKey accountId: Int?): SimpleOpportunity {
        return opportunityMapperExt.findOpportunityAssoWithConvertedLead(leadId!!)
    }

    companion object {

        init {
            ClassInfoMap.put(OpportunityServiceImpl::class.java, ClassInfo(ModuleNameConstants.CRM, CrmTypeConstants.OPPORTUNITY))
        }
    }
}
